apps/frontend/core-dapp/src/stores/BalancerStore.ts

```diff
@@ -0,0 +1,131 @@
+import { BigNumber, utils } from 'ethers'
+import { SupportedNetworks } from 'prepo-constants'
+import { getContractAddress } from 'prepo-utils'
+import { makeAutoObservable, reaction, runInAction } from 'mobx'
+import { RootStore } from './RootStore'
+import { SupportedContracts } from '../lib/contract.types'
+import { BalancerQueriesAbi__factory } from '../../generated/typechain'
+import { supportedContracts } from '../lib/supported-contracts'
+
+const wethAddressByNetowrk: Partial<Record<SupportedNetworks, string>> = {
+  arbitrumOne: '0x82aF49447D8a07e3bd95BD0d56f35241523fBab1',
+}
+
+const WETH_DECIMALS = 18
+
+const wstEthWethBalancerPoolIdByNetwork: Partial<Record<SupportedNetworks, string>> = {
+  arbitrumOne: '0x36bf227d6bac96e2ab1ebb5492ecec69c691943f000200000000000000000316',
+}
+
+export class BalancerStore {
+  private readonly balancerQueriesInterface = BalancerQueriesAbi__factory.createInterface()
+  private readonly root: RootStore
+  private wstEthPriceInEthBN: BigNumber | undefined = undefined
+
+  constructor(root: RootStore) {
+    this.root = root
+
+    makeAutoObservable(this)
+
+    reaction(
+      () => root.web3Store.blockNumber,
+      async (blockNumber) => {
+        const { wstEthPriceInEthBN } = this
+
+        if (
+          wstEthPriceInEthBN !== undefined &&
+          blockNumber !== undefined &&
+          // Only query every 90 blocks (~30s)
+          blockNumber % 90 !== 0
+        )
+          return
```

Comment:
Wdyt if we rely on timestamp instead? I monitored how fast blocks are processed on arbitrum for awhile, it doesn't seem so stable sometimes. It basically generates 30-40 blocks every 1-2 seconds, sometimes more/less for short/longer interval

Because of that, I feel like it's better to use a timestamp to control this, for example on our Web3Store, if you check the frequency of handleNewBlock, it's actually being triggered 30-40 times every second (can check this by console.log in the function), but if you inspect the network request, we're only sending 1 request every second, this is because we've implemented a 1sec stall timeout when creating the provider.

I reckon we could use the same approach as provider which is to fire a call only if the previous call was fired more than 1sec ago

So I'd probably do something like this (I've tested this timestamp approach locally btw): 

```suggestion
      async () => {
        const now = new Date().getTime()
        if (this.lastCalled + SEC_IN_MS > now) return
        this.lastCalled = now
```

Reply:
Also since there's still chance of getting rate limited, I'd use this timestamp approach to add a cool off period if an error is thrown, I'll provide example in the next comment

Reply:
Changed it to a timestamp approach every 30s. 3s seems excessively short, I don't think we need that much accuracy for displaying prices.

Reply:
I think a shorter refresh is better tho. This lets large balances see their balance ticking every few seconds which is more exciting. Since the calls on Multicall are handled properly now (on average 4sec per call), I think we can use 3s here. I had the same configuration run locally for over 10mins without getting any rate limit error.

Reply:
> This lets large balances see their balance ticking every few seconds which is more exciting

Counterpoint: ticking numbers give me anxiety :) but I don't have a strong opinion here either way